lib/recursive-methods.rb

```diff
@@ -1,49 +1,119 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n < 0
+    raise ArgumentError
+  end
+  return 1 if n <= 1
+  return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  #web help source: https://medium.com/@txea/reverse-a-string-with-recursion-a3f8505ffb73
+  reversed_str = ""
+  i = 0
+  while i < s.length
+    reversed_str = s[i] + reversed_str
+    i += 1
+  end
+  reversed_str
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def reverse_inplace(s)
```

Comment:
This works, but it's not in place and time complexity and space complexity are O(n<sup>2</sup>)